### ISSUE TITLE
Database migration error with move from 0.25.0 -> 0.26.1 : table 'collection' already exists

### ISSUE THREAD

**Issue by lesiki** [metabase/metabase#6075]
Searched through open & closed issues but didn't see anyone facing migration problems on 0.26.1.

I'm aware there was a db migraiton issue #6050 which required 0.26.1 to be released, but I'm facing this when I went straight to 0.26.1 from 0.25.0:

```
09-29 14:00:20 ESC[1mINFO metabase.coreESC[0m :: Starting Metabase version v0.26.1 (dacff52 release-0.26.1) ...
09-29 14:00:20 ESC[1mINFO metabase.coreESC[0m :: System timezone is 'Etc/UTC' ...
09-29 14:00:21 ESC[1mWARN metabase.driverESC[0m :: No -init-driver function found for 'metabase.driver.google'
09-29 14:00:21 ESC[1mINFO metabase.coreESC[0m :: Setting up and migrating Metabase DB. Please sit tight, this may take a minute...
09-29 14:00:21 ESC[1mINFO metabase.dbESC[0m :: ESC[36mVerifying mysql Database Connection ...ESC[0m
09-29 14:00:22 ESC[1mINFO metabase.dbESC[0m :: Verify Database Connection ...  <U+2705>
09-29 14:00:22 ESC[1mINFO metabase.dbESC[0m :: Running Database Migrations...
09-29 14:00:22 ESC[1mINFO metabase.dbESC[0m :: Setting up Liquibase...
09-29 14:00:22 ESC[1mINFO metabase.dbESC[0m :: Liquibase is ready.
09-29 14:00:22 ESC[1mINFO metabase.dbESC[0m :: Checking if Database has unrun migrations...
09-29 14:00:26 ESC[1mINFO metabase.dbESC[0m :: Database has unrun migrations. Waiting for migration lock to be cleared...
09-29 14:00:26 ESC[1mINFO metabase.dbESC[0m :: Migration lock is cleared. Running migrations...
com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Table 'collection' already exists
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
        at com.mysql.jdbc.Util.handleNewInstance(Util.java:404)
        at com.mysql.jdbc.Util.getInstance(Util.java:387)
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:942)
        at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3966)
        at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3902)
        at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2526)
        at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2673)
        at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2549)
        at com.mysql.jdbc.PreparedStatement.executeInternal(PreparedStatement.java:1861)
        at com.mysql.jdbc.PreparedStatement.executeUpdateInternal(PreparedStatement.java:2073)
        at com.mysql.jdbc.PreparedStatement.executeUpdateInternal(PreparedStatement.java:2009)
        at com.mysql.jdbc.PreparedStatement.executeLargeUpdate(PreparedStatement.java:5098)
        at com.mysql.jdbc.PreparedStatement.executeUpdate(PreparedStatement.java:1994)
        at clojure.java.jdbc$db_do_execute_prepared_statement$fn__10886.invoke(jdbc.clj:901)
        at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:748)
        at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:684)
        at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:697)
        at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:684)
        at clojure.java.jdbc$db_do_execute_prepared_statement.invokeStatic(jdbc.clj:900)
        at clojure.java.jdbc$db_do_execute_prepared_statement.invoke(jdbc.clj:894)
        at clojure.java.jdbc$db_do_prepared.invokeStatic(jdbc.clj:932)
        at clojure.java.jdbc$db_do_prepared.invoke(jdbc.clj:912)
        at clojure.java.jdbc$execute_BANG_$execute_helper__10945.invoke(jdbc.clj:1161)
        at clojure.java.jdbc$execute_BANG_.invokeStatic(jdbc.clj:1163)
        at clojure.java.jdbc$execute_BANG_.invoke(jdbc.clj:1145)
        at clojure.java.jdbc$execute_BANG_.invokeStatic(jdbc.clj:1156)
        at clojure.java.jdbc$execute_BANG_.invoke(jdbc.clj:1145)
        at metabase.db$migrate_up_if_needed_BANG_.invokeStatic(db.clj:164)
        at metabase.db$migrate_up_if_needed_BANG_.invoke(db.clj:151)
        at metabase.db$migrate_BANG_$fn__19651.invoke(db.clj:247)
        at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:714)
        at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:684)
        at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:740)
        at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:684)
        at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:697)
        at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:684)
        at metabase.db$migrate_BANG_.invokeStatic(db.clj:235)
        at metabase.db$migrate_BANG_.invoke(db.clj:218)
        at metabase.db$run_schema_migrations_BANG_.invokeStatic(db.clj:365)
        at metabase.db$run_schema_migrations_BANG_.invoke(db.clj:360)
        at metabase.db$setup_db_BANG_.invokeStatic(db.clj:383)
        at metabase.db$setup_db_BANG_.doInvoke(db.clj:376)
        at clojure.lang.RestFn.invoke(RestFn.java:421)
        at metabase.core$init_BANG_.invokeStatic(core.clj:103)
        at metabase.core$init_BANG_.invoke(core.clj:82)
        at metabase.core$start_normally.invokeStatic(core.clj:187)
        at metabase.core$start_normally.invoke(core.clj:181)
        at metabase.core$_main.invokeStatic(core.clj:285)
        at metabase.core$_main.doInvoke(core.clj:280)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at metabase.core.main(Unknown Source)
09-29 14:00:43 ESC[1mERROR metabase.coreESC[0m :: Metabase Initialization FAILED:  Table 'collection' already exists
09-29 14:00:43 ESC[1mINFO metabase.coreESC[0m :: Metabase Shutting Down ...
09-29 14:00:43 ESC[1mINFO metabase.coreESC[0m :: Metabase Shutdown COMPLETE
```

My setup:
- Metabase is using MySQL 5.5.49-0ubuntu0.14.04.1 (Ubuntu) for local metbase db, and same for remote database

The next steps I did:
1) Tried to run again, but got error stating the DATABASECHANGELOGLOCK was active.
2) Tried the `java -jar metabase.jar migrate release-locks` which exited without error, but changelog lock was still active
3) Manually released changeloglock (I'm familiar with liquibase so did this direct on my mysql db)
4) Tried to run again, faced same `Table 'collection' already exists`
5) Tried `java -jar metabase.jar migrate force`, which exits without error, but re-run still gives same "table exists" error on first attempt, and db migration lock error for repeat attempts

Latest two migrations run in my db:
```
*************************** 1. row ***************************
           ID: 46
       AUTHOR: camsaul
     FILENAME: migrations/000_migrations.yaml
 DATEEXECUTED: 2017-09-29 13:18:28
ORDEREXECUTED: 45
     EXECTYPE: EXECUTED
       MD5SUM: 7:aab12e940225b458986e15cf53d5d816
  DESCRIPTION: addNotNullConstraint columnName=row, tableName=report_dashboardcard; addNotNullConstraint columnName=col, tableName=report_dashboardcard; addDefaultValue columnName=row, tableName=report_dashboardcard; addDefaultValue columnName=col, tableName=rep...
     COMMENTS:
          TAG: NULL
    LIQUIBASE: 3.5.3
     CONTEXTS: NULL
       LABELS: NULL
DEPLOYMENT_ID: 6691093466
*************************** 2. row ***************************
           ID: 45
       AUTHOR: tlrobinson
     FILENAME: migrations/000_migrations.yaml
 DATEEXECUTED: 2016-10-13 06:52:56
ORDEREXECUTED: 44
     EXECTYPE: EXECUTED
       MD5SUM: 7:9198081e3329df7903d9016804ef0cf0
  DESCRIPTION: addColumn tableName=report_dashboardcard; addNotNullConstraint columnName=visualization_settings, tableName=report_dashboardcard
     COMMENTS:
          TAG: NULL
    LIQUIBASE: 3.5.2
     CONTEXTS: NULL
       LABELS: NULL
DEPLOYMENT_ID: 6341573081
2 rows in set (0.00 sec)
```

**Comment by salsakran** [comment]
You might try upgrading through the intermediate versions - 
http://downloads.metabase.com/v0.25.1/metabase.jar
http://downloads.metabase.com/v0.25.2/metabase.jar
and then going to 0.26.1

If that works, please let us know. If it doesn't ... we'll need to dig in further.

**Comment by lesiki** [comment]
When I try 0.25.1 I get the same migration failure:
```
com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Table 'collection' already exists
```

**Comment by senior** [comment]
@lesiki There isn't a migration named collection. What I think is happening is that it's attempting to run a migration that includes the creation of the `collection` table, but you've already got a collection table, so it's failing. What's odd is that the collection table was included long before 0.25.1 (it's included in 0.22.0).

My best guess now is that your migrations table is in an odd state. Can you run the below query and paste in the results?

```sql
select id,md5sum from databasechangelog
```

**Comment by lesiki** [comment]
@senior Ah, I assumed (from looking at the description field of other past migrations) that the changelog entry which added `collection` would have that word somewhere in the description field.

The query results:

```
mysql> select id,md5sum from databasechangelog;
ERROR 1146 (42S02): Table 'metabase.databasechangelog' doesn't exist
# included this just in case case sensitivity is an issue here? I assume not.

mysql> select id,md5sum from DATABASECHANGELOG;
+----+------------------------------------+
| id | md5sum                             |
+----+------------------------------------+
| 1  | 7:1b0742414a26a9b52ed5622ce7516ad7 |
| 2  | 7:e63cf4e800ec122c1a19c04b8c07cc3f |
| 4  | 7:1ed887e91a846f4d6cbe84d1efd126c4 |
| 5  | 7:593149128c8f3a7e1f37a483bc67a924 |
| 6  | 7:d24f2f950306f150d87c4208520661d5 |
| 7  | 7:baec0ec600ccc9bdadc176c1c4b29b77 |
| 8  | 7:ea2727c7ce666178cff436549f81ddbd |
| 9  | 7:c05cf8a25248b38e281e8e85de4275a2 |
| 10 | 7:97fec69516d0dfe424ea7365f51bb87e |
| 11 | 7:c7ef8b4f4dcb3528f9282b51aea5bb2a |
| 12 | 7:f78e18f669d7c9e6d06c63ea9929391f |
| 13 | 7:f27286894439bef33ff93761f9b32bc4 |
| 14 | 7:2bc8ad9e775091db38a79ee610eafc97 |
| 15 | 7:50c72a51651af76928c06f21c9e04f97 |
| 16 | 7:70484ee18e0067988963d3042343e300 |
| 17 | 7:5401ec35a5bd1275f93a7cac1ddd7591 |
| 18 | 7:329d897d44ba9893fdafc9ce7e876d73 |
| 19 | 7:e8fa976811e4d58d42a45804affa1d07 |
| 20 | 7:e9f9ab6a3f7a54d3cf1ab70bf7f288b9 |
| 21 | 7:10f47e47d4a6b5479702485355968d60 |
| 22 | 7:cb6776ec86ab0ad9e74806a5460b9085 |
| 23 | 7:43b9662bd798db391d4bbb7d4615bf0d |
| 24 | 7:b77dea78a3645ab993b30d949abee28a |
| 25 | 7:c40bd959d6a76be99e324a8a821c0bde |
| 26 | 7:ac7f40d2a3fbf3fea7936aa79bb1532b |
| 27 | 7:6daeade421b85d930b07f95d8d6a14c5 |
| 28 | 7:335e7e6b32dcbeb392150b3c3db2d5eb |
| 29 | 7:7b0bb8fcb7de2aa29ce57b32baf9ff31 |
| 30 | 7:7b5245de5d964eedb5cd6fdf5afdb6fd |
| 31 | 7:347281cdb65a285b03aeaf77cb28e618 |
| 32 | 7:ff40b5fbe06dc5221d0b9223992ece25 |
| 32 | 7:143cf1041ee698943cc6d1c6650105b4 |
| 34 | 7:e65d70b4c914cfdf5b3ef9927565e899 |
| 35 | 7:ab80b6b8e6dfc3fa3e8fa5954e3a8ec4 |
| 36 | 7:de534c871471b400d70ee29122f23847 |
| 37 | 7:487dd1fa57af0f25edf3265ed9899588 |
| 38 | 7:5e32fa14a0c34b99027e25901b7e3255 |
| 39 | 7:a63ada256c44684d2649b8f3c28a3023 |
| 40 | 7:0ba56822308957969bf5ad5ea8ee6707 |
| 41 | 7:e1aa5b70f61426b29d74d38936e560de |
| 42 | 7:779407e2ea3b8d89092fc9f72e29fdaa |
| 43 | 7:dbc18c8ca697fc335869f0ed0eb5f4cb |
| 44 | 7:1d09a61933bbc5a01b0ddef7bd1b1336 |
| 45 | 7:9198081e3329df7903d9016804ef0cf0 |
| 46 | 7:aab12e940225b458986e15cf53d5d816 |
+----+------------------------------------+
45 rows in set (0.00 sec)
```

**Comment by senior** [comment]
@lesiki I think you've got a database that is not in a consistent state. The md5sums from the most recent migrations are correct. With the results you posted above, it looks like the last successful migration that ran was around version 0.21.0. According to the above, migration 47 which was first included in 0.22.0, hasn't ran. The collection table, was included in migration 47, so it shouldn't be in your database.

How long have you been running 0.25.0? I would have expected that version to fail in the same was 0.26.1 failed. Are there rows in your collections table (i.e. `select count(*) from COLLECTIONS`)? Could the Metabase database be coexisting with another application that has a collections table? I'm also wondering if subsequent migrations have ran as well, do you have a `COLLECTION_REVISION` table (added in migration 48)?

@camsaul Were there any bugs in the ~ 0.22.0 time frame that would cause this migration inconsistency? I didn't see anything in the issue tracker, but am curious if that would explain some of this.

**Comment by camsaul** [comment]
@senior I went through. The issue tracker as well and didn't see anything. I'm not sure what the root cause of this issue is.

`java -jar metabase.jar migrate force` might be able to fix this. It executes the migration DDL statements but ignores one that fail for some reason or another (like the `CREATE TABLE collection`) and continues rather than dumping the Exception and quitting. So you can give that a try. I'd recommend making a backup of your DB before you do so just to be safe.

**Comment by lesiki** [comment]
@senior thanks for giving this so much attention.

> How long have you been running 0.25.0? I would have expected that version to fail in the same was 0.26.1 failed.

I'm now starting to wonder if we somehow switched between versions at some point, and if that could somehow affect migrations. I can't see how it could - my experience of liquibase is that you can happily skip versions and still have smooth migration, and if for any reason you go backwards, it just means the app crashes when the schema isn't as expected, but it wouldn't mess up the structure. To answer the question - I believe we installed 0.25.0 just as soon as it was released (we get the update emails), and had no issues on that version.

> Are there rows in your collections table (i.e. select count(*) from COLLECTIONS)?

No, it's empty

> Could the Metabase database be coexisting with another application that has a collections table?

No, metabase is the only thing using mysql on that server, the database with our data is on another server.

> I'm also wondering if subsequent migrations have ran as well, do you have a COLLECTION_REVISION table (added in migration 48)?

Yes, it exists and is also empty.

@camsaul thanks for looking at this as well. As mentioned above I actually tried `migrate force` - it gave me a bunch of 'success' statements, but running the app after that still results in same outcome.

Some experiments I'll try and report back on:
1) Go back to 0.25.0 and see if the app starts up ok.
2) (After performing db backup) try dropping these tables and see how the migration responds - in theory even though the state of run migrations doesn't add up, I could catch back up to 0.26.1 by allowing it to perform all the missing migrations? Of course there may well be others after the `create table collection` that would fail.

**Comment by lesiki** [comment]
I've managed to get my db & migrations back in sync.

I reviewed the liquibase migrations in the 0.26.1 tag, and saw that I was about 15 migrations behind according to the databasechangelog table. The errors being reported on startup was that some migrations were missing even though their schema changes had already been done, so I compared the changes introduced in each migration (50 - 66).

I noticed an interesting pattern: where there were `addTable`s, my database already had the table. Where there were `addColumn`, the new columns did _not_ already exist in my db. So whatever caused this inconsistent state only introduced new empty tables, but didn't introduce new columns and indices. There was one 'createTable' command in the migrations where my db did not have the table already existing: `dimension` table from migration no 58 was not present. 58 was present in 0.25.0 (which is what I believe we were running before my upgrade), so this only adds to the mystery for me..

I deleted the tables that were in the db but shouldn't be, and with that got my installation working again - launched the app and the migrations happily ran the whole way through. 

Hoping this behaviour is additional info that could help come up with hypotheses. It's a stretch, but one thought I had is that if the underlying ORM/db driver has a 'create-table-if-not-found' mode (I know hibernate does), then if we accidentally hot-swapped the jar file with the app running, we might have triggered creation of these tables at runtime, which then causes this migration issue on subsequent restart?

As far as my personal issue goes I'm unblocked, thanks loads for the input @senior, @salsakran and @camsaul . Let me know if there's any more I can share in case other come across this.
